### PR TITLE
V1.0

### PR DESCRIPTION
**Pull request by quebramadeira**
@yegorgeorgiyev  comment me
@nikita-safonov comment me
@Shlepanets comment me
@Cakemanator comment me

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,38 @@
...
+Docker compose installed:
+  - sudo curl -L "https://github.com/docker/compose/releases/download/1.29.2/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
+  - sudo chmod +x /usr/local/bin/docker-compose
+
+## quick start
+Place to the home directory:
+  - config folder with all the contents;
+  - headnode folder with all the contents;
+  - worker folder with all the contents;
+  - hadoop-3.1.2.tar.gz file from https://hadoop.apache.org/release/3.1.2.html
```
Comment (Cakemanator) [inline README.md]: Better to use a link https://archive.apache.org/dist/hadoop/common/hadoop-3.1.2/hadoop-3.1.2.tar.gz

File: README.md [hunk 2]
```diff
@@ -1 +1,38 @@
...
+Create docker images from Dockerfiles:
+  - docker build -t headnode:1.0 -f headnode/Dockerfile .
+  - docker build -t worker:1.0 -f worker/Dockerfile .
+
+Run headnode and worker containers using:
+  - docker run -p 8088:8088 -p 9870:9870 --add-host=headnode:192.168.10.2 --add-host=worker:192.168.10.3 --network=hadoop --ip=192.168.10.2 -itd <headnode_image_id>
+  - docker run --add-host=headnode:192.168.10.2 --add-host=worker:192.168.10.3 --network=hadoop --ip=192.168.10.3 -itd <worker_image_id>
+
+Or using docker-compose.yml file:
+  - docker-compose up -d
```
Comment (Cakemanator) [inline README.md:36]: image
Created network:
image

File: headnode/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+FROM centos
+MAINTAINER Alexander Nemchinov <quantumrods@gmail.com>
```
Comment (Cakemanator) [inline headnode/Dockerfile:2]: Please use more best practices for Dockerfiles

File: worker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,30 @@
+FROM centos
+MAINTAINER Alexander Nemchinov <quantumrods@gmail.com>
```
Comment (Cakemanator) [inline worker/Dockerfile:2]: Please use more best practices for Dockerfiles

**Review by Cakemanator — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (quebramadeira) [reply README.md:36]: I have used this line for network creation, everything works fine, even with a new VM:
`docker network create --gateway 192.168.10.1 --subnet 192.168.10.0/24 hadoop`

File: headnode/Dockerfile (hunk 3 above)
Comment (quebramadeira) [reply headnode/Dockerfile:2]: `FROM centos:8`

File: worker/Dockerfile (hunk 4 above)
Comment (quebramadeira) [reply worker/Dockerfile:2]: `FROM centos:8`

File: headnode/Dockerfile (hunk 3 above)
Comment (Cakemanator) [reply headnode/Dockerfile:2]: What about optimization of repeated instructions (RUN, COPY, etc)?

File: worker/Dockerfile (hunk 4 above)
Comment (Cakemanator) [reply worker/Dockerfile:2]: What about optimization of repeated instructions (RUN, COPY, etc)?

**Review by Cakemanator — CHANGES_REQUESTED** [review]

File: headnode/Dockerfile (hunk 3 above)
Comment (quebramadeira) [reply headnode/Dockerfile:2]: FROM centos:8
MAINTAINER Alexander Nemchinov <[quantumrods@gmail.com](mailto:quantumrods@gmail.com)>
WORKDIR /etc/yum.repos.d/
RUN sed -i 's/mirrorlist/#mirrorlist/g' /etc/yum.repos.d/CentOS-* \
        && sed -i 's|#baseurl=http://mirror.centos.org|baseurl=http://vault.centos.org|g' /etc/yum.repos.d/CentOS-* \
        && yum -y update && yum -y remove java && yum install -y  java-1.8.0-openjdk
WORKDIR /opt
COPY hadoop-3.1.2.tar.gz /opt
RUN tar -xf hadoop-3.1.2.tar.gz && mkdir /usr/local/hadoop \
        &&  ln -s /opt/hadoop-3.1.2/ /usr/local/hadoop/current
COPY config/hadoop-env.sh config/core-site.xml config/hdfs-site.xml config/yarn-site.xml /usr/local/hadoop/current/etc/hadoop/
WORKDIR /usr/local/hadoop/current/etc/hadoop
RUN sed -i 's:%PATH_TO_OPENJDK8_INSTALLATION%:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.312.b07-2.el8_5.x86_64/jre:g' hadoop-env.sh \
        && sed -i 's:%PATH_TO_HADOOP_INSTALLATION:/usr/local/hadoop/current:g' hadoop-env.sh \
        && sed -i 's:%HADOOP_HEAP_SIZE%:512M:g' hadoop-env.sh \
        && sed -i 's:%HDFS_NAMENODE_HOSTNAME%:headnode:g' core-site.xml \
        && sed -i 's:%NAMENODE_DIRS%:/opt/mount1/namenode-dir,/opt/mount2/namenode-dir:g' hdfs-site.xml \
        && sed -i 's:%DATANODE_DIRS%:/opt/mount1/datanode-dir,/opt/mount2/datanode-dir:g' hdfs-site.xml \
        && sed -i 's:%YARN_RESOURCE_MANAGER_HOSTNAME%:headnode:g' yarn-site.xml \
        && sed -i 's:%NODE_MANAGER_LOCAL_DIR%:/opt/mount1/nodemanager-local-dir,/opt/mount2/nodemanager-local-dir:g' yarn-site.xml \
        && sed -i 's:%NODE_MANAGER_LOG_DIR%:/opt/mount1/nodemanager-log-dir,/opt/mount2/nodemanager-log-dir:g' yarn-site.xml
ENV HADOOP_HOME="/usr/local/hadoop/current"
RUN groupadd -g 4000 hadoop && useradd -g hadoop -u 1500 hadoop \
        && useradd -g hadoop -u 1600 yarn \
        && useradd -g hadoop -u 1700 hdfs \
        && mkdir /opt/hadoop-3.1.2/logs \
        && chown :hadoop /opt/hadoop-3.1.2/logs \
        && setfacl -d -R -m g:hadoop:rwx /opt/hadoop-3.1.2/logs \
        && setfacl -R -m g:hadoop:rwx /opt/hadoop-3.1.2/logs \
        && mkdir -p /opt/mount1/namenode-dir \
        && mkdir -p /opt/mount2/namenode-dir \
        && chown hdfs /opt/mount1/namenode-dir \
        && chown hdfs /opt/mount2/namenode-dir \
        && chown :hadoop /opt/mount1/namenode-dir \
    && chown :hadoop /opt/mount2/namenode-dir
VOLUME /opt/mount1 opt/mount2
COPY headnode/runheadnode.sh /runheadnode.sh
RUN chmod a+x /runheadnode.sh
CMD ["/runheadnode.sh"]

**Pull request merged**